micromasters_import/management/commands/queries/008_import_combined_final_grade.sql

```diff
@@ -0,0 +1,106 @@
+----This script is to migrate the grades for course run and exam from MicroMaster:
+----  It computes score from FinalGrade and ProctoredExamGrade using formula - course run grade * 0.4 + exam grade * 0.6
+----  grades need to exist in both FinalGrade (course run) and ProctoredExamGrade (exam)
+----  For 'passed' field, true if both have passed= true, otherwise false.
+
+WITH courses_have_exam AS (
+    SELECT
+        DISTINCT mm_examrun.course_id
+    FROM micromasters.exams_examrun AS mm_examrun
+    JOIN micromasters.courses_course AS mm_course
+        ON mm_examrun.course_id = mm_course.id
+),
+ --- this is to compute the best grade learner gets for each course run regardless if passed or not
+best_courserun_grades AS (
+    SELECT
+        mm_grade.user_id,
+        mm_grade.course_run_id,
+        MAX(mm_grade.grade) AS max_grade
+    FROM micromasters.grades_finalgrade AS mm_grade
+    JOIN micromasters.courses_courserun AS mm_courserun
+        ON mm_grade.course_run_id = mm_courserun.id
+    JOIN courses_have_exam
+        ON mm_courserun.course_id = courses_have_exam.course_id
+    WHERE mm_grade.status = 'complete'
+    GROUP BY mm_grade.user_id, mm_grade.course_run_id
+),
+--- this is to compute the best exam grades learner get for every course run in the exam they signed up for
+best_exam_grades AS (
+    SELECT
+        mm_exam_grade.user_id,
+        mm_exam_grade.course_id,
+        mm_courserun.id AS course_run_id,
+        MAX(mm_exam_grade.score) AS max_score
+    FROM micromasters.grades_proctoredexamgrade AS mm_exam_grade
+    JOIN micromasters.courses_courserun AS mm_courserun
+       ON mm_exam_grade.course_id = mm_courserun.course_id
+    JOIN courses_have_exam
+       ON mm_exam_grade.course_id = courses_have_exam.course_id
+    GROUP BY mm_exam_grade.user_id, mm_courserun.id, mm_exam_grade.course_id
+),
+combined_final_grades AS (
+    SELECT
+        best_grade.user_id,
+        best_grade.course_run_id,
```

Comment:
Are we capturing two different course_run_ids? I see `best_grade.course_run_id` here and then `mm_courserun.course_id` below. Sorry my SQL is so weak, I'm not clear about what the source of these two are, nor how they might be used.

Reply:
`best_grade.course_run_id` comes from [grades_finalgrade](https://micromasters-rc.odl.mit.edu/admin/grades/finalgrade/1/change/) . In order to match MicroMaster course_run_id with MITxOnline, we need the corresponding _edx_course_key_ so it can be inserted into MITxOnline. it's like this:
`MM grades_finalgrade.course_run_id => MM courses_courserun.edx_course_key => MO courses_courserun.courseware_id => MO.courses_courserun.id`